File: sdks/javascript/fastdatabroker_multitenancy.ts
/**
 * FastDataBroker TypeScript/JavaScript SDK - Multi-Tenant Client Library
 * Version 0.1.16 - With Multi-Tenancy Support
 */

import * as fs from 'fs';
import * as path from 'path';

export const Version = '0.1.16';

/**
 * Tenant Configuration
 */
export interface TenantConfig {
  tenant_id: string;
  tenant_name: string;
  api_key_prefix: string;
  rate_limit_rps: number;
  max_connections: number;
  max_message_size: number;
  retention_days: number;
  enabled?: boolean;
  metadata?: Record<string, any>;
}

/**
 * Server Configuration
 */
export interface ServerConfig {
  bind_address: string;
  port: number;
  enable_tls: boolean;
  cert_path: string;
  key_path: string;
}

/**
 * Application Configuration
 */
export interface AppConfig {
  name: string;
  version: string;
  environment: string;
}

/**
 * Application Settings
 */
export interface AppSettings {
  app: AppConfig;
  server: ServerConfig;
  tenants: TenantConfig[];
}

/**
 * Utility class for configuration management
 */
export class ConfigurationManager {
  /**
   * Load configuration from JSON file with environment overrides
   */
  static loadFromFile(filePath: string, environment: string = 'development'): AppSettings {
    if (!fs.existsSync(filePath)) {
      throw new Error(`Configuration file not found: ${filePath}`);
    }

    const data = fs.readFileSync(filePath, 'utf-8');
    const baseSettings: AppSettings = JSON.parse(data);

    // Try to load environment-specific config
    const dir = path.dirname(filePath);
    const fileName = path.basename(filePath);
    const nameWithoutExt = path.basename(fileName, path.extname(fileName));
    const envFile = path.join(dir, `${nameWithoutExt}.${environment}.json`);

    if (fs.existsSync(envFile)) {
      const envData = fs.readFileSync(envFile, 'utf-8');
      const envSettings: AppSettings = JSON.parse(envData);

      if (envSettings.app) {
        baseSettings.app = envSettings.app;
      }
      if (envSettings.server) {
        baseSettings.server = envSettings.server;
      }
      if (envSettings.tenants && envSettings.tenants.length > 0) {
        for (const tenant of envSettings.tenants) {
          const found = baseSettings.tenants.some(t => t.tenant_id === tenant.tenant_id);
          if (!found) {
            baseSettings.tenants.push(tenant);
          }
        }
      }
    }

    return baseSettings;
  }

  /**
   * Get tenant by ID
   */
  static getTenant(settings: AppSettings, tenantId: string): TenantConfig | undefined {
    return settings.tenants.find(t => t.tenant_id === tenantId);
  }

  /**
   * Get tenant by API key prefix
   */
  static getTenantByApiKey(settings: AppSettings, apiKey: string): TenantConfig | undefined {
    return settings.tenants.find(t => apiKey.startsWith(t.api_key_prefix));
  }

  /**
   * Validate tenant configuration
   */
  static validateTenant(tenant: TenantConfig): void {
    if (!tenant.tenant_id) {
      throw new Error('TenantId cannot be empty');
    }
    if (!tenant.api_key_prefix || !tenant.api_key_prefix.endsWith('_')) {
      throw new Error('ApiKeyPrefix must end with "_"');
    }
    if (tenant.rate_limit_rps <= 0) {
      throw new Error('RateLimitRps must be greater than 0');
    }
    if (tenant.max_connections <= 0) {
      throw new Error('MaxConnections must be greater than 0');
    }
  }
}

/**
 * Priority levels for messages
 */
export enum Priority {
  Deferred = 50,
  Normal = 100,
  High = 150,
  Urgent = 200,
  Critical = 255
}

/**
 * Notification delivery channels
 */
export enum NotificationChannel {
  Email = 'email',
  WebSocket = 'websocket',
  Push = 'push',
  Webhook = 'webhook'
}

/**
 * Push notification platforms
 */
export enum PushPlatform {
  Firebase = 'firebase',
  APNs = 'apns',
  FCM = 'fcm',
  WebPush = 'webpush'
}

/**
 * Message envelope for FastDataBroker
 */
export class Message {
  tenantId: string;
  senderId: string;
  recipientIds: string[];
  subject: string;
  content: Buffer | Uint8Array;
  priority: Priority = Priority.Normal;
  ttlSeconds?: number;
  tags: Map<string, string>;
  requireConfirm: boolean = false;

  constructor(
    tenantId: string,
    senderId: string,
    recipientIds: string[],
    subject: string,
    content: Buffer | Uint8Array
  ) {
    this.tenantId = tenantId;
    this.senderId = senderId;
    this.recipientIds = recipientIds || [];
    this.subject = subject;
    this.content = content || new Buffer(0);
    this.tags = new Map();
  }
}

/**
 * Delivery result for a sent message
 */
export interface DeliveryResult {
  messageId: string;
  tenantId: string;
  status: string;
  deliveredChannels: number;
  details: Record<string, any>;
}

/**
 * WebSocket client information
 */
export interface WebSocketClientInfo {
  clientId: string;
  userId: string;
  tenantId: string;
  connectedAt: Date;
}

/**
 * Webhook configuration
 */
export interface WebhookConfig {
  url: string;
  headers?: Record<string, string>;
  retries?: number;
  timeoutMs?: number;
  verifySsl?: boolean;
}

/**
 * Multi-Tenant FastDataBroker client
 */
export class Client {
  private host: string;
  private port: number;
  private tenantId: string;
  private apiKey: string;
  private settings: AppSettings;
  private connected: boolean = false;
  private wsClients: Map<string, WebSocketClientInfo>;

  /**
   * Create a new client
   */
  constructor(host: string = 'localhost', port: number = 6379) {
    this.host = host;
    this.port = port;
    this.tenantId = '';
    this.apiKey = '';
    this.settings = { app: {}, server: {}, tenants: [] } as AppSettings;
    this.wsClients = new Map();
  }

  /**
   * Create a tenant-aware client
   */
  static createWithTenant(
    tenantId: string,
    apiKey: string,
    host: string = 'localhost',
    port: number = 6379
  ): Client {
    if (!tenantId) {
      throw new Error('TenantId cannot be empty');
    }
    if (!apiKey) {
      throw new Error('ApiKey cannot be empty');
    }

    const client = new Client(host, port);
    client.tenantId = tenantId;
    client.apiKey = apiKey;
    return client;
  }

  /**
   * Create a client from AppSettings with tenant validation
   */
  static createFromSettings(
    settings: AppSettings,
    tenantId: string,
    apiKey: string
  ): Client {
    if (!settings) {
      throw new Error('Settings cannot be null');
    }
    if (!tenantId) {
      throw new Error('TenantId cannot be empty');
    }
    if (!apiKey) {
      throw new Error('ApiKey cannot be empty');
    }

    const tenant = ConfigurationManager.getTenant(settings, tenantId);
    if (!tenant) {
      throw new Error(`Tenant '${tenantId}' not found in configuration`);
    }

    if (!apiKey.startsWith(tenant.api_key_prefix)) {
      throw new Error(`API key does not match tenant prefix: ${tenant.api_key_prefix}`);
    }

    const client = new Client(settings.server.bind_address, settings.server.port);
    client.tenantId = tenantId;
    client.apiKey = apiKey;
    client.settings = settings;
    return client;
  }

  /**
   * Connect to FastDataBroker server with tenant context
   */
  async connect(): Promise<boolean> {
    try {
      if (!this.tenantId) {
        throw new Error('TenantId must be set before connecting');
      }

      this.connected = true;
      console.log(`[TENANT: ${this.tenantId}] Connected to FastDataBroker at ${this.host}:${this.port}`);
      return true;
    } catch (error) {
      console.error(`[TENANT: ${this.tenantId}] Connection failed: ${error}`);
      return false;
    }
  }

  /**
   * Send a message with tenant isolation
   */
  async sendMessage(message: Message): Promise<DeliveryResult> {
    if (!this.connected) {
      throw new Error('Not connected. Call connect() first.');
    }

    if (!message) {
      throw new Error('Message cannot be null');
    }

    if (!message.tenantId) {
      message.tenantId = this.tenantId;
    }

    if (message.tenantId !== this.tenantId) {
      throw new Error('Message tenant does not match client tenant');
    }

    return {
      messageId: this.generateId(),
      tenantId: this.tenantId,
      status: 'success',
      deliveredChannels: 1,
      details: {}
    };
  }

  /**
   * Register WebSocket client (tenant-isolated)
   */
  registerWebSocketClient(clientId: string, userId: string): boolean {
    if (!clientId || !userId) {
      return false;
    }

    const clientInfo: WebSocketClientInfo = {
      clientId,
      userId,
      tenantId: this.tenantId,
      connectedAt: new Date()
    };

    this.wsClients.set(clientId, clientInfo);
    return true;
  }

  /**
   * Unregister WebSocket client
   */
  unregisterWebSocketClient(clientId: string): boolean {
    return this.wsClients.delete(clientId);
  }

  /**
   * Register webhook endpoint (tenant-isolated)
   */
  registerWebhook(channel: NotificationChannel, config: WebhookConfig): boolean {
    if (!config || !config.url) {
      return false;
    }
    return true;
  }

  /**
   * Generate API key for a client (tenant-aware)
   */
  generateApiKey(clientId: string): string {
    if (!clientId) {
      throw new Error('ClientId cannot be empty');
    }

    const tenant = ConfigurationManager.getTenant(this.settings, this.tenantId);
    if (!tenant) {
      throw new Error(`Tenant '${this.tenantId}' not found`);
    }

    const guid = this.generateId();
    return `${tenant.api_key_prefix}${guid}`;
  }

  /**
   * Get current tenant configuration
   */
  getTenantConfig(): TenantConfig | undefined {
    return ConfigurationManager.getTenant(this.settings, this.tenantId);
  }

  /**
   * Disconnect
   */
  disconnect(): void {
    this.connected = false;
    console.log(`[TENANT: ${this.tenantId}] Disconnected from FastDataBroker`);
  }

  /**
   * Private helper to generate IDs
   */
  private generateId(): string {
    return `${Date.now()}${Math.random().toString(36).substring(2, 15)}`;
  }

  /**
   * Get connection status
   */
  isConnected(): boolean {
    return this.connected;
  }

  /**
   * Get tenant ID
   */
  getTenantId(): string {
    return this.tenantId;
  }
}

/**
 * Factory function to create client from config file
 */
export async function createClient(
  configPath: string,
  tenantId: string,
  apiKey: string,
  environment: string = 'development'
): Promise<Client> {
  const settings = ConfigurationManager.loadFromFile(configPath, environment);
  return Client.createFromSettings(settings, tenantId, apiKey);
}

File: sdks/go/multitenancy/fastdatabroker_multitenancy.go
package multitenancy

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const MultiTenancyVersion = "0.1.16"

// Priority represents message priority
type Priority uint8

const (
	PriorityDeferred Priority = 50
	PriorityNormal   Priority = 100
	PriorityHigh     Priority = 150
	PriorityUrgent   Priority = 200
	PriorityCritical Priority = 255
)

// NotificationChannel represents delivery channels
type NotificationChannel int

const (
	ChannelEmail NotificationChannel = iota
	ChannelWebSocket
	ChannelPush
	ChannelWebhook
)

// TenantConfig represents tenant-specific configuration
type TenantConfig struct {
	TenantID       string                 `json:"tenant_id"`
	TenantName     string                 `json:"tenant_name"`
	APIKeyPrefix   string                 `json:"api_key_prefix"`
	RateLimitRps   uint32                 `json:"rate_limit_rps"`
	MaxConnections uint32                 `json:"max_connections"`
	MaxMessageSize uint64                 `json:"max_message_size"`
	RetentionDays  uint32                 `json:"retention_days"`
	Enabled        bool                   `json:"enabled"`
	Metadata       map[string]interface{} `json:"metadata,omitempty"`
}

// Validate checks if tenant configuration is valid
func (tc *TenantConfig) Validate() error {
	if tc.TenantID == "" {
		return fmt.Errorf("TenantID cannot be empty")
	}
	if tc.APIKeyPrefix == "" || !strings.HasSuffix(tc.APIKeyPrefix, "_") {
		return fmt.Errorf("APIKeyPrefix must end with '_'")
	}
	if tc.RateLimitRps == 0 {
		return fmt.Errorf("RateLimitRps must be greater than 0")
	}
	if tc.MaxConnections == 0 {
		return fmt.Errorf("MaxConnections must be greater than 0")
	}
	return nil
}

// ServerConfig represents server configuration
type ServerConfig struct {
	BindAddress string `json:"bind_address"`
	Port        uint16 `json:"port"`
	EnableTLS   bool   `json:"enable_tls"`
	CertPath    string `json:"cert_path"`
	KeyPath     string `json:"key_path"`
}

// AppConfig represents application configuration
type AppConfig struct {
	Name        string `json:"name"`
	Version     string `json:"version"`
	Environment string `json:"environment"`
}

// AppSettings represents complete application settings
type AppSettings struct {
	App     AppConfig      `json:"app"`
	Server  ServerConfig   `json:"server"`
	Tenants []TenantConfig `json:"tenants"`
}

// LoadFromFile loads AppSettings from JSON file with environment overrides
func LoadFromFile(filePath string, environment string) (*AppSettings, error) {
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return nil, fmt.Errorf("configuration file not found: %s", filePath)
	}

	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read config file: %w", err)
	}

	var settings AppSettings
	if err := json.Unmarshal(data, &settings); err != nil {
		return nil, fmt.Errorf("failed to parse config file: %w", err)
	}

	// Try to load environment-specific config
	envFile := filepath.Join(
		filepath.Dir(filePath),
		strings.TrimSuffix(filepath.Base(filePath), filepath.Ext(filePath))+"."+environment+".json",
	)

	if _, err := os.Stat(envFile); err == nil {
		envData, err := os.ReadFile(envFile)
		if err == nil {
			var envSettings AppSettings
			if err := json.Unmarshal(envData, &envSettings); err == nil {
				if envSettings.App != (AppConfig{}) {
					settings.App = envSettings.App
				}
				if envSettings.Server != (ServerConfig{}) {
					settings.Server = envSettings.Server
				}
				if len(envSettings.Tenants) > 0 {
					for _, tenant := range envSettings.Tenants {
						found := false
						for _, existing := range settings.Tenants {
							if existing.TenantID == tenant.TenantID {
								found = true
								break
							}
						}
						if !found {
							settings.Tenants = append(settings.Tenants, tenant)
						}
					}
				}
			}
		}
	}

	return &settings, nil
}

// GetTenant retrieves tenant by ID
func (as *AppSettings) GetTenant(tenantID string) *TenantConfig {
	for i := range as.Tenants {
		if as.Tenants[i].TenantID == tenantID {
			return &as.Tenants[i]
		}
	}
	return nil
}

// GetTenantByAPIKey retrieves tenant by API key prefix
func (as *AppSettings) GetTenantByAPIKey(apiKey string) *TenantConfig {
	for i := range as.Tenants {
		if strings.HasPrefix(apiKey, as.Tenants[i].APIKeyPrefix) {
			return &as.Tenants[i]
		}
	}
	return nil
}

// PushPlatform represents push notification platforms
type PushPlatform int

const (
	Firebase PushPlatform = iota
	APNs
	FCM
	WebPush
)

// Message represents a FastDataBroker message
type Message struct {
	TenantID       string
	SenderID       string
	RecipientIDs   []string
	Subject        string
	Content        []byte
	Priority       Priority
	TTLSeconds     *int64
	Tags           map[string]string
	RequireConfirm bool
}

// DeliveryResult represents message delivery result
type DeliveryResult struct {
	MessageID         string
	TenantID          string
	Status            string
	DeliveredChannels int
	Details           map[string]interface{}
}

// WebSocketClientInfo represents a WebSocket client connection
type WebSocketClientInfo struct {
	ClientID    string
	UserID      string
	TenantID    string
	ConnectedAt time.Time
}

// WebhookConfig represents webhook configuration
type WebhookConfig struct {
	URL       string
	Headers   map[string]string
	Retries   int
	TimeoutMs int
	VerifySSL bool
}

// Client represents a multi-tenant FastDataBroker client
type Client struct {
	Host      string
	Port      int
	TenantID  string
	APIKey    string
	Settings  *AppSettings
	connected bool
	wsClients map[string]*WebSocketClientInfo
}

// NewClient creates a new FastDataBroker client
func NewClient(host string, port int) *Client {
	return &Client{
		Host:      host,
		Port:      port,
		connected: false,
		wsClients: make(map[string]*WebSocketClientInfo),
		Settings:  &AppSettings{},
	}
}

// NewClientWithTenant creates a new multi-tenant aware client
func NewClientWithTenant(tenantID string, apiKey string, host string, port int) (*Client, error) {
	if tenantID == "" {
		return nil, fmt.Errorf("TenantID cannot be empty")
	}
	if apiKey == "" {
		return nil, fmt.Errorf("APIKey cannot be empty")
	}

	return &Client{
		Host:      host,
		Port:      port,
		TenantID:  tenantID,
		APIKey:    apiKey,
		connected: false,
		wsClients: make(map[string]*WebSocketClientInfo),
		Settings:  &AppSettings{},
	}, nil
}

// NewClientFromSettings creates a client from AppSettings with tenant validation
func NewClientFromSettings(settings *AppSettings, tenantID string, apiKey string) (*Client, error) {
	if settings == nil {
		return nil, fmt.Errorf("settings cannot be nil")
	}
	if tenantID == "" {
		return nil, fmt.Errorf("TenantID cannot be empty")
	}
	if apiKey == "" {
		return nil, fmt.Errorf("APIKey cannot be empty")
	}

	// Validate tenant exists
	tenant := settings.GetTenant(tenantID)
	if tenant == nil {
		return nil, fmt.Errorf("tenant '%s' not found in configuration", tenantID)
	}

	// Validate API key prefix
	if !strings.HasPrefix(apiKey, tenant.APIKeyPrefix) {
		return nil, fmt.Errorf("API key does not match tenant prefix: %s", tenant.APIKeyPrefix)
	}

	return &Client{
		Host:      settings.Server.BindAddress,
		Port:      int(settings.Server.Port),
		TenantID:  tenantID,
		APIKey:    apiKey,
		Settings:  settings,
		connected: false,
		wsClients: make(map[string]*WebSocketClientInfo),
	}, nil
}

// Connect establishes connection with tenant context
func (c *Client) Connect() error {
	if c.TenantID == "" {
		return fmt.Errorf("TenantID must be set before connecting")
	}

	c.connected = true
	fmt.Printf("[TENANT: %s] Connected to FastDataBroker at %s:%d\n", c.TenantID, c.Host, c.Port)
	return nil
}

// SendMessage sends a message with tenant isolation
func (c *Client) SendMessage(msg *Message) (*DeliveryResult, error) {
	if !c.connected {
		return nil, fmt.Errorf("not connected. Call Connect first")
	}

	if msg == nil {
		return nil, fmt.Errorf("message cannot be nil")
	}

	if msg.TenantID == "" {
		msg.TenantID = c.TenantID
	}

	if msg.TenantID != c.TenantID {
		return nil, fmt.Errorf("message tenant does not match client tenant")
	}

	return &DeliveryResult{
		MessageID:         generateID(),
		TenantID:          c.TenantID,
		Status:            "success",
		DeliveredChannels: 1,
		Details:           make(map[string]interface{}),
	}, nil
}

// RegisterWebSocketClient registers a WebSocket client (tenant-isolated)
func (c *Client) RegisterWebSocketClient(clientID string, userID string) bool {
	if clientID == "" || userID == "" {
		return false
	}

	c.wsClients[clientID] = &WebSocketClientInfo{
		ClientID:    clientID,
		UserID:      userID,
		TenantID:    c.TenantID,
		ConnectedAt: time.Now().UTC(),
	}

	return true
}

// UnregisterWebSocketClient unregisters a WebSocket client
func (c *Client) UnregisterWebSocketClient(clientID string) bool {
	if _, exists := c.wsClients[clientID]; exists {
		delete(c.wsClients, clientID)
		return true
	}
	return false
}

// RegisterWebhook registers a webhook endpoint (tenant-isolated)
func (c *Client) RegisterWebhook(channel NotificationChannel, config *WebhookConfig) bool {
	if config == nil || config.URL == "" {
		return false
	}
	return true
}

// GenerateAPIKey generates API key for a client (tenant-aware)
func (c *Client) GenerateAPIKey(clientID string) (string, error) {
	if clientID == "" {
		return "", fmt.Errorf("ClientID cannot be empty")
	}

	tenant := c.Settings.GetTenant(c.TenantID)
	if tenant == nil {
		return "", fmt.Errorf("tenant '%s' not found", c.TenantID)
	}

	return tenant.APIKeyPrefix + generateID()[:16], nil
}

// GetTenantConfig returns current tenant configuration
func (c *Client) GetTenantConfig() *TenantConfig {
	return c.Settings.GetTenant(c.TenantID)
}

// Disconnect closes the connection
func (c *Client) Disconnect() {
	c.connected = false
	fmt.Printf("[TENANT: %s] Disconnected\n", c.TenantID)
}

// CreateClient helper function to load config and create client
func CreateClient(configPath string, tenantID string, apiKey string, environment string) (*Client, error) {
	settings, err := LoadFromFile(configPath, environment)
	if err != nil {
		return nil, err
	}

	return NewClientFromSettings(settings, tenantID, apiKey)
}

// Helper function to generate IDs
func generateID() string {
	return fmt.Sprintf("%d%d", time.Now().UnixNano(), time.Now().Nanosecond())
}
